noisy-simulator/src/density_matrix_simulator.rs

```diff
@@ -0,0 +1,346 @@
+        self.check_out_of_bounds_qubits(qubits)?;
+
+        let mut tmp_state = self.state.clone()?;
+        apply_kernel(
+            &mut tmp_state.data,
+            instrument.total_effect_transposed(),
+            qubits,
+        )?;
+        let total_effect_trace = tmp_state.trace();
+        if total_effect_trace < TOLERANCE {
+            let err = Error::ProbabilityZeroEvent;
+            handle_error!(self, err);
+        }
+        let mut last_non_zero_trace_outcome: usize = 0;
+        let mut last_non_zero_trace: f64 = 0.0;
+        let mut summed_probability: f64 = 0.0;
+
+        for outcome in 0..instrument.num_operations() {
+            if summed_probability > random_sample {
+                break;
+            }
+            tmp_state = self.state.clone()?;
+            apply_kernel(
+                &mut tmp_state.data,
+                instrument.operation(outcome).effect_matrix_transpose(),
+                qubits,
+            )?;
+            let outcome_trace = tmp_state.trace();
+            summed_probability += outcome_trace / total_effect_trace;
+            if outcome_trace >= TOLERANCE {
+                last_non_zero_trace_outcome = outcome;
+                last_non_zero_trace = outcome_trace;
+            }
+        }
+
+        if summed_probability + TOLERANCE <= random_sample || last_non_zero_trace < TOLERANCE {
+            let err = Error::FailedToSampleInstrumentOutcome;
+            handle_error!(self, err);
+        }
+
+        if let Err(err) = self.state.as_mut()?.apply_operation_matrix(
+            instrument.operation(last_non_zero_trace_outcome).matrix(),
+            qubits,
+        ) {
+            handle_error!(self, err);
+        };
+        if let Err(err) = self
+            .state
+            .as_mut()?
+            .renormalize_with_trace(last_non_zero_trace)
+        {
+            handle_error!(self, err);
+        };
+        Ok(last_non_zero_trace_outcome)
+    }
+
+    /// Returns the `DensityMatrix` if the simulator is in a valid state.
+    fn state(&self) -> Result<&DensityMatrix, &Error> {
+        self.state.as_ref()
+    }
+
+    /// Set state of the quantum system.
+    fn set_state(&mut self, new_state: DensityMatrix) -> Result<(), Error> {
+        if self.dimension != new_state.dimension() {
+            return Err(Error::InvalidState(format!(
+                "the provided state should have the same dimensions as the quantum system's state, {} != {}",
+                self.dimension,
+                new_state.dimension(),
+            )));
+        }
+        if !new_state.is_normalized() {
+            return Err(Error::InvalidState(format!(
+                "`state` is not normalized, trace is {}",
+                new_state.trace()
+            )));
+        }
+        if !new_state.is_hermitian() {
+            return Err(Error::InvalidState("`state` is not Hermitian".to_string()));
+        }
+        self.state = Ok(new_state);
```

Comment:
Why is state a result?

Reply:
This is an intentional design choice. The simulator state becomes invalid on some cases (when things go wrong numerically). The alternative is to have state be a `DensityMatrix` and have another field called `last_error` that is an `Option<Error>`. But this is prone to errors, because you could accidentally have `last_error` be Some(...) after the simulator has recovered.